src/sidechain.cpp

```diff
@@ -0,0 +1,466 @@
+        uint16_t counter;
+        s >> counter;
+        if (change < 0 && !counter) return;  // may be surprising if change is <-1
+        counter += change;
+        s.clear();
+        s << counter;
+    });
+}
+
+uint256 CalculateDrivechainWithdrawBlindedHash(const CTransaction& tx) {
+    CMutableTransaction mtx(tx);
+    mtx.vin[0].SetNull();
+    mtx.vout[0].SetNull();
+    return mtx.GetHash();
+}
+
+uint256 CalculateDrivechainWithdrawInternalHash(const uint256& blinded_hash, const uint8_t sidechain_id) {
+    // Internally, we hash the bundle_hash with the sidechain_id to avoid conflicts between sidechains
+    uint256 internal_hash;
+    CSHA256().Write(blinded_hash.data(), blinded_hash.size()).Write(&sidechain_id, sizeof(sidechain_id)).Finalize(internal_hash.data());
+    return internal_hash;
+}
+
+bool UpdateDrivechains(const CTransaction& tx, CCoinsViewCache& view, CTxUndo &txundo, int block_height, BlockValidationState& state)
+{
+    Assert(tx.IsCoinBase());
+
+    std::vector<unsigned char> sidechain_proposal_list, withdraw_proposal_list;
+    std::set<uint8_t> saw_withdraw_proposed_for_sidechain;
+    bool proposed_a_sidechain{false}, saw_sidechain_acks{false};
+
+    for (auto& out : tx.vout) {
+        if (out.scriptPubKey.size() < 5) continue;
+        if (out.scriptPubKey[0] != OP_RETURN) continue;
+        // FIXME: The rest should probably be serialised, but neither BIP300 nor its reference implementation does that
+        static constexpr uint8_t BIP300_HEADER_SIDECHAIN_PROPOSE[] = {0xd5, 0xe0, 0xc4, 0xaf};  // n.b. 20 sigops
+        static constexpr uint8_t BIP300_HEADER_SIDECHAIN_ACK[]     = {0xd6, 0xe1, 0xc5, 0xbf};
+        static constexpr uint8_t BIP300_HEADER_WITHDRAW_PROPOSE[]  = {0xd4, 0x5a, 0xa9, 0x43};  // n.b. 67 byte push followed by only 32 bytes
+        static constexpr uint8_t BIP300_HEADER_WITHDRAW_ACK[]      = {0xd7, 0x7d, 0x17, 0x76};  // n.b. 23-byte push followed by variable bytes
+        if (std::equal(&out.scriptPubKey[1], &out.scriptPubKey[5], BIP300_HEADER_WITHDRAW_ACK)) {
+            const uint8_t data_format = out.scriptPubKey[6];
+            // TODO: Implement formats 3+? Or at least validate
+                // NOTE data_format 2 changed to 0 FIXME
+                // TODO: (new) data format 2 sets it to the ACKs from the previous block - but those aren't known, have the same cost, and encourages blind upvoting; so can we get rid of it?
+                // TODO: data format 3 upvotes any bundle leading its rivals by at least 50 ACKs -- also encourages blind upvoting :/
+            // TODO: How is vote vector actually encoded?
+            // TODO: Block is invalid if there are no bundles proposed at all
+            // FIXME: Presumably blocks should only be able to vote once - this is missing in the BIP
+            for (int sidechain_id = 0; sidechain_id < 0x100; ++sidechain_id) {
+                // FIXME: bounds checking
```

Comment:
probably something like `out.scriptPubKey.size() >= 6 + (0x100 * data_format)`  before we start counting votes